stablehlo/dialect/TypeInference.cpp

```diff
@@ -948,6 +948,63 @@ LogicalResult inferSortOp(
   return success();
 }
 
+LogicalResult inferTransposeOp(Optional<Location> loc, Value operand,
+                               DenseIntElementsAttr permutation,
+                               SmallVectorImpl<Type>& inferredReturnTypes) {
+  auto type = operand.getType();
+  auto rankedTy = type.dyn_cast<RankedTensorType>();
+  if (!rankedTy) {
+    inferredReturnTypes.emplace_back(type);
+    return success();
+  }
+  int64_t rank = rankedTy.getRank();
+  if (permutation.getType().getRank() != 1)
+    return emitOptionalError(loc, "TransposeOp permutation has rank ",
+                             permutation.getType().getRank(),
+                             " instead of rank 1");
+
+  if (permutation.size() != rank)
+    return emitOptionalError(loc, "TransposeOp operand rank ", rank,
+                             " does not match permutation size ",
+                             permutation.size());
+
+  std::vector<int64_t> range(rank);
+  std::iota(range.begin(), range.end(), 0);
+  if (!std::is_permutation(range.begin(), range.end(), permutation.begin()))
+    return emitOptionalError(loc,
+                             "attribute permutation must be a permutation"
+                             " of [",
+                             range, "] but got ", permutation);
+
+  llvm::SmallVector<int64_t, 4> inputBounds(rank, ShapedType::kDynamicSize);
+  bool hasBounds = false;
+  BoundedDialectInterface* dialect = nullptr;
```

Comment:
We'll really need to cut down on the boilerplate around getting and setting bounds, but let's wait until similar logic is added to other ops and then see what we can deduplicate.

Reply:
Sure makes sense.